File: 004/Haskell/Problem004.hs
module Problem004 (findPalindromeProduct, isPalindrome) where

findPalindromeProduct :: Integer -> Integer
findPalindromeProduct n =
  maximum [x * y | x <- range, y <- range, isPalindrome (x * y)]
    where range = [n..(n*10-1)]

isPalindrome n = reverse (show n) == (show n)

File: 004/Erlang/problem004.erl
-module(problem004).
-export([findPalindromeProduct/1, isPalindrome/1]).
-use(lists).

findPalindromeProduct(N) -> 
  lists:max([X * Y || X <- range(N), Y <- range(N), isPalindrome(X*Y)]).
  
range(N) -> lists:seq(N, N*10-1).

isPalindrome(N) -> lists:reverse(integer_to_list(N)) =:= integer_to_list(N).
